src/UTF8.cs

```diff
@@ -1388,20 +1388,28 @@ private unsafe static (int utfadjust, int scalaradjust) calculateErrorPathadjust
                             prevIncomplete = Vector128<byte>.Zero;
                             // Often, we have a lot of ASCII characters in a row.
                             int localasciirun = 16;
-                            if (processedLength + localasciirun + 64 <= inputLength)
+                            if (processedLength + localasciirun + 16 <= inputLength)
                             {
-                                for (; processedLength + localasciirun + 64 <= inputLength; localasciirun += 64)
+                                Vector128<byte> block = AdvSimd.LoadVector128(pInputBuffer + processedLength + localasciirun);
+                                if (AdvSimd.Arm64.MaxAcross(Vector128.AsUInt32(AdvSimd.And(block, v80))).ToScalar() == 0)
                                 {
-                                    Vector128<byte> block1 = AdvSimd.LoadVector128(pInputBuffer + processedLength + localasciirun);
-                                    Vector128<byte> block2 = AdvSimd.LoadVector128(pInputBuffer + processedLength + localasciirun + 16);
-                                    Vector128<byte> block3 = AdvSimd.LoadVector128(pInputBuffer + processedLength + localasciirun + 32);
-                                    Vector128<byte> block4 = AdvSimd.LoadVector128(pInputBuffer + processedLength + localasciirun + 48);
-                                    Vector128<byte> or = AdvSimd.Or(AdvSimd.Or(block1, block2), AdvSimd.Or(block3, block4));
-                                    if (AdvSimd.Arm64.MaxAcross(or).ToScalar() > 127)
+                                    localasciirun += 16;
+                                    for (; processedLength + localasciirun + 64 <= inputLength; localasciirun += 64)
                                     {
-                                        break;
+                                        Vector128<byte> block1 = AdvSimd.LoadVector128(pInputBuffer + processedLength + localasciirun);
+                                        Vector128<byte> block2 = AdvSimd.LoadVector128(pInputBuffer + processedLength + localasciirun + 16);
+                                        Vector128<byte> block3 = AdvSimd.LoadVector128(pInputBuffer + processedLength + localasciirun + 32);
+                                        Vector128<byte> block4 = AdvSimd.LoadVector128(pInputBuffer + processedLength + localasciirun + 48);
+                                        Vector128<byte> or = AdvSimd.Or(AdvSimd.Or(block1, block2), AdvSimd.Or(block3, block4));
```

Comment:
It appears that CSE does not perform as well as it could here, preventing load coalescing. Note that in this specific case using platform intrinsic does not have a difference with plain `Vector128.Load`:
```cs
byte* pAsciiBuffer = pInputBuffer + asciirun;
Vector128<byte> block1 = Vector128.Load(pAsciiBuffer);
Vector128<byte> block2 = Vector128.Load(pAsciiBuffer + 16);
Vector128<byte> block3 = Vector128.Load(pAsciiBuffer + 32);
Vector128<byte> block4 = Vector128.Load(pAsciiBuffer + 48);
```
Before:
```asm
G_M146_IG06:  ;; offset=0x0048
            sxtw    x5, w4
            ldr     q16, [x0, x5]
            add     x6, x0, x5
            ldr     q17, [x6, #0x10]
            orr     v16.16b, v16.16b, v17.16b
            add     x6, x0, x5
            ldr     q17, [x6, #0x20]
            add     x5, x0, x5
            ldr     q18, [x5, #0x30]
            orr     v17.16b, v17.16b, v18.16b
            orr     v16.16b, v16.16b, v17.16b
            umaxv   b16, v16.16b
            umov    w5, v16.b[0]
            cmp     w5, #127
            bgt     G_M146_IG07
            add     w4, w4, #64
            add     w5, w4, #64
            cmp     w5, w1
            ble     G_M146_IG06
```
After:
```asm
G_M146_IG06:  ;; offset=0x0048
            add     x5, x0, w4, SXTW
            ldp     q16, q17, [x5]
            orr     v16.16b, v16.16b, v17.16b
            ldp     q17, q18, [x5, #0x20]
            orr     v17.16b, v17.16b, v18.16b
            orr     v16.16b, v16.16b, v17.16b
            umaxv   b16, v16.16b
            umov    w5, v16.b[0]
            cmp     w5, #127
            bgt     G_M146_IG07
            add     w4, w4, #64
            add     w5, w4, #64
            cmp     w5, w1
            ble     G_M146_IG06
```

Does not matter much probably but one step away from perfect codegen :)
cc @EgorBo

Reply:
@neon-sunset Oddly enough, on my Apple M2, I am getting better results (consistently) with the code as it is... If you have ARM hardware, can you try running...

```
 dotnet run --configuration Release --filter "*Latin*"
```

I am getting rock solid better results with the current code. 

I have not looked at the generated code.

Reply:
> Does not matter much probably

Turns out this assumption was completely wrong - with the change now hits 100 GiB/s mark as measured by the benchmark suite.
```
BenchmarkDotNet v0.13.12, macOS 15.0 (24A5264n) [Darwin 24.0.0]
Apple M1 Pro, 1 CPU, 8 logical and 8 physical cores
.NET SDK 9.0.100-preview.6.24314.10
  [Host]     : .NET 9.0.0 (9.0.24.30702), Arm64 RyuJIT AdvSIMD
  Job-WPCVMR : .NET 9.0.0 (9.0.24.30702), Arm64 RyuJIT AdvSIMD

MaxIterationCount=10  MinIterationCount=2  WarmupCount=1 
```
Baseline:
| Method                              | FileName                   | Mean     | Error     | StdDev    | Speed (GB/s) |
|------------------------------------ |--------------------------- |---------:|----------:|----------:|------------- |
| SIMDUtf8ValidationRealData          | data/Latin-Lipsum.utf8.txt | 1.026 us | 0.0076 us | 0.0012 us | 84.75        |
| DotnetRuntimeUtf8ValidationRealData | data/Latin-Lipsum.utf8.txt | 2.451 us | 0.0310 us | 0.0017 us | 35.47        |

After:
| Method                              | FileName                   | Mean       | Error    | StdDev  | Speed (GB/s) |
|------------------------------------ |--------------------------- |-----------:|---------:|--------:|------------- |
| SIMDUtf8ValidationRealData          | data/Latin-Lipsum.utf8.txt |   868.0 ns |  5.39 ns | 0.83 ns | 100.16       |
| DotnetRuntimeUtf8ValidationRealData | data/Latin-Lipsum.utf8.txt | 2,459.9 ns | 37.83 ns | 5.85 ns | 35.34        |

Reply:
Are you game to do a PR?

I see you are using .NET 9, this could be the difference?

Reply:
> Are you game to do a PR?

Sure, thanks! Will look through implementation to see if there is anything else that stands out.

For the Latin scenario, .NET 9 here shouldn't make *significant* difference, there were changes regarding loop handling but here the win is purely via merging the loads and removing redundant adds (that might also add a false dependency?) .

Reply:
> Sure, thanks! Will look through implementation to see if there is anything else that stands out.

Please do!!!

Reply:
> For the Latin scenario, .NET 9 here shouldn't make significant difference, there were changes regarding loop handling but here the win is purely via merging the loads and removing redundant adds (that might also add a false dependency?) .

I installed .NET 9 and I see no difference. Your proposal is still a net negative. But I am going to toy a bit more with it.

Reply:
> I installed .NET 9 and I see no difference. Your proposal is still a net negative. But I am going to toy a bit more with it.

Is it regressing performance on 8cx gen3 and/or Neoverse N1?

Reply:
> I assume it should be a similar improvement on M2...

I would assume so but it seems that it is not the case.